README
I only added support for the latest (2.23) SagePay protocol.
Addon page: http://addons.oscommerce.com/info/441


ChangeLog.txt
26 Oct 2009
-----------
sagepay_form-v1.20.2

updated to use sage pay protocol 2.23

3 June 2009
-----------
sagepay_form-v1.20.1

updated English language file as well, omitted from last update.

14 May 2009
-----------
sagepay_form-v1.19.1

All references to Protx have been removed and replaced with Sagepay within the code and all filenames.

protx --> sagepay
Protx --> Sagepay
PROTX --> SAGEPAY

Corrected minor spelling mistake and error in installation instructions

note: Spanish and German versions are still in English... these should be translated prior to use.


14 May 2009
-----------
sagepay_form-v1.19
URL modifications as a result of PROTX becoming SAGEPAY

Note added to the installation instructions about compy name change as there are many remaining references to Protx in the code.
Modified live and test addresses to point to new Sagepay URLs instead of Protx. 
Removed the commented out references to defunct Protx URLs

...Further work is required to tidy up coding to remove all reference to Protx but this is not essential from a functional perspective.


6th October 2008
----------------
protx_form-v1.18.2
decrypt_protx.php line 88 changed from
    $Crypt = $_REQUEST['Crypt'];
to
    $Crypt = $_REQUEST['crypt'];
so that the submitted data is in a field called crypt lower case

24 Feb 2008
-----------
protx_form-v1.18.1
Removed a .= that had accidentally crept in.

24 Feb 2008
-----------
protx_form-v1.18.1
Adjusted the handling of adding items to the Basket variable so as to restrict
its size to 7495 characters. Protx v2.22 states maximum size of this field is
7500 characters.

Fixed issue with before_process() not returning the $Status variable.

Consumed the change suggested by vitaliy on Aug 7, 2007.

06 Aug 2007
-----------
protx_form-v1.17 - Added AUTHENTICATE AND AUTHORISE return code also removed previous comment marks.

Please see osCommerce forum thread for details of code changes
http://forums.oscommerce.com/index.php?showtopic=272122

If it needs fixed, well.....ehhhmmmm?

03 Aug 2007
-----------
protx_form-V1.16 - Change to VendorTxCode to produce extra uniqueness
*all suggestions courtesy of theINtoy*

The changes to the code here are minor, but to take care of any issues on non-uniqueness
1.  The store name is now taken from the PROTX form as this has to be unique
		(oscommerce stores have the potential to be named the same)

2.  The customers ID has been added to the end of this string,so that it will reduce the potential 
		for 2 seperate customers buying something at EXACTLY the same second.
		Now their ID added will make the extra difference required

Line 130 - added $customer_id to the global line to fetch it.
Changed     global $order, $currencies, $currency;
to	    global $order, $currencies, $currency, $customer_id;

Line 142 - changed the STORE_NAME reference to be the same as PROTX STORE NAME, added $customer_id after the date/time
		this will keep the references in order still as it is only the customers ID at the end for uniqueness.
Changed      $plain = "VendorTxCode=" . STORE_NAME . date('YmdHis') . "&";
to           $plain = "VendorTxCode=" . MODULE_PAYMENT_PROTX_FORM_VENDOR_NAME . date('YmdHis') . $customer_id . "&";

NOTE: You could change this to include some kind of dividers between each part of the reference 
seperating "store name_datetime_customerid"
e.g.  $plain = "VendorTxCode=" . MODULE_PAYMENT_PROTX_FORM_VENDOR_NAME ."_" . date('YmdHis') . "_" . $customer_id . "&";


02 Aug 2007
-----------
protx_form-V1.15 - Fix for VendorTxCode to add store name automatically

Changed the VendorTxCode so that it now adds your store name to the VendorTxCode automatically
to make your code unique as suggested by theINtoy.

Line 142
Changed       $plain = "VendorTxCode=" . date('YmdHis') . "&";
to	      $plain = "VendorTxCode=" . STORE_NAME . date('YmdHis') . "&";


01 Aug 2007
------------
protx_form-V1.14 - Fix for VendorTxCode
Changed the VendorTxCode so that it now uses a 24 hour clock for the hours. 
Changed       $plain = "VendorTxCode=" . date('Ymdhis') . "&";
to            $plain = "VendorTxCode=" . date('YmdHis') . "&";

NB. I personally believe that the VendoTxCode should have a unique identifier
either prepended or appended to this date to make it unique in the protx
system. As things stand the VendorTxCode has a chance of not being unique in
within Protx if 2 protx clients attempt to submit at the same second. This
could happen.
01 Aug 2007
------------
protx_form-v1.13 Added global $order on line 17 and changed submittion URL's for updated Protx ones

Line 17 global $order added

Line 31-39 were changed to NEW URL's also added a line to include a reference to the OLD URLs

Line 31-39 used to be
      if (MODULE_PAYMENT_PROTX_FORM_TEST_STATUS == 'true') {
        $this->form_action_url = 'https://ukvpstest.protx.com/vps2form/submit.asp';
		//$this->form_action_url = 'https://ukvpstest.protx.com/showpost/showpost.asp';
		//$this->form_action_url = 'https://ukvpstest.protx.com/VSPSimulator/VSPFormGateway.asp';
	 //	$this->form_action_url = '../../decrypt_protx.php';
      } else {
        $this->form_action_url = 'https://ukvps.protx.com/vps2form/submit.asp';
      }
    }


Line 31-42 now is
      if (MODULE_PAYMENT_PROTX_FORM_TEST_STATUS == 'true') {
        $this->form_action_url = 'https://ukvpstest.protx.com/vspgateway/service/vspform-register.vsp';
        //old LIVE URL was https://ukvpstest.protx.com/vps2form/submit.asp
		//$this->form_action_url = 'https://ukvpstest.protx.com/showpost/showpost.asp';
		//$this->form_action_url = 'https://ukvpstest.protx.com/VSPSimulator/VSPFormGateway.asp';
        //$this->form_action_url = '../../decrypt_protx.php';

      } else {
        $this->form_action_url = 'https://ukvps.protx.com/vspgateway/service/vspform-register.vsp ';
        //old LIVE URL was https://ukvps.protx.com/vps2form/submit.asp
      }
    }


25 July 2007
------------

catalog/decrypt_sagepay.php
<html>
<head>
<title>DE-CRYPT SAGEPAY</title>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
</head>

<body>
<?
/*
decrypt_sagepay.php is not needed except for de-bugging and does not have to be uploaded to the live site.

To use decrypt_sagepay.php navigate to line 87 which should look like this:

$de_crypt = SimpleXor(base64_decode($Crypt),'geavj7yt4G6Wiy8x');

the string 'geavj7yt4G6Wiy8x' is a fake SAGEPAY password.

You must substitute yours (the same one you enter in the admin panel) here for decrypt_sagepay.php to work.

then go to around line 30 in sagepay_form.php

            if (MODULE_PAYMENT_SAGEPAY_FORM_TEST_STATUS == 'true') {
        $this->form_action_url = 'https://test.sagepay.com/vps2form/submit.asp';
        //$this->form_action_url = '../../decrypt_sagepay.php';

and uncomment the  line $this->form_action_url = '../../decrypt_sagepay.php';
*/


// Set $toBrowser to false if you want to log this information rather than showing it on a browser page.
// If logging is enabled, a file called sagepay.log with be created in your docRoot.
// ** For extra security, please remember to delete sagepay.log once you have finished debugging **

// Set $toBrowser to true if you want to show this information on a browser page.
$toBrowser = true;

function simpleXor($InString, $Key) {
      $KeyList = array();
      $output = "";

      for($i = 0; $i < strlen($Key); $i++){
        $KeyList[$i] = ord(substr($Key, $i, 1));
      }

      for($i = 0; $i < strlen($InString); $i++) {
        $output.= chr(ord(substr($InString, $i, 1)) ^ ($KeyList[$i % strlen($Key)]));
      }

      return $output;
}

    function getToken($thisString) {

      $Tokens = array("Status","StatusDetail","VendorTxCode","VPSTxID","TxAuthNo","Amount","AVSCV2");

      $output = array();
      $resultArray = array();

      for ($i = count($Tokens)-1; $i >= 0 ; $i--){
        $start = strpos($thisString, $Tokens[$i]);
        if ($start !== false){
          $resultArray[$i]->start = $start;
          $resultArray[$i]->token = $Tokens[$i];
        }
      }

      sort($resultArray);

      for ($i = 0; $i<count($resultArray); $i++){
        $valueStart = $resultArray[$i]->start + strlen($resultArray[$i]->token) + 1;
        if ($i==(count($resultArray)-1)) {
          $output[$resultArray[$i]->token] = substr($thisString, $valueStart);
        } else {
          $valueLength = $resultArray[$i+1]->start - $resultArray[$i]->start - strlen($resultArray[$i]->token) - 2;
          $output[$resultArray[$i]->token] = substr($thisString, $valueStart, $valueLength);
        }
      }

      return $output;
    }
// ** retrieve the information posted from the previous form. **
$ThisVendorTxCode = $_REQUEST['VendorTxCode'];
$VPSProtocol = $_REQUEST['VPSProtocol'];
$TxType = $_REQUEST['TxType'];
$Vendor = $_REQUEST['Vendor'];
$Crypt = $_REQUEST['crypt'];

$outStr = $ThisVendorTxCode."<BR>" . $VPSProtocol."<BR>" . $TxType."<BR>" . $Vendor."<BR>";

$de_crypt = SimpleXor(base64_decode($Crypt),'geavj7yt4G6Wiy8x');

 $resultArray = getToken($de_crypt);

       for ($i = 0; $i<count($resultArray); $i++){
        echo $resultArray[$i]."<BR>";
      }

$outStr .= $de_crypt."<BR>";

if ( $toBrowser )
{
    echo $outStr;
}
else
{
    error_log( $Crypt . "\n\n", 3, "./sagepay.log");
    error_log( nl2br( $outStr ) , 3, "./sagepay.log");
    echo "<h2>Payment Service Not Currently Available</h2><br>Please use the back button and select a different service.<br>Sorry.</h2>";
}
?>
</body>
</html>


catalog/includes/modules/payment/sagepay_form.php
<?php
/*
  $Id$

  osCommerce, Open Source E-Commerce Solutions
  http://www.oscommerce.com

  Copyright (c) 2004 osCommerce

  Released under the GNU General Public License
*/

  class sagepay_form {
    var $code, $title, $description, $enabled;

// class constructor
    function sagepay_form() {
      global $order;
      $this->code = 'sagepay_form';
      $this->title = MODULE_PAYMENT_SAGEPAY_FORM_TEXT_TITLE;
      $this->description = MODULE_PAYMENT_SAGEPAY_FORM_TEXT_DESCRIPTION;
      $this->sort_order = MODULE_PAYMENT_SAGEPAY_FORM_SORT_ORDER;
      $this->enabled = ((MODULE_PAYMENT_SAGEPAY_FORM_STATUS == 'True') ? true : false);

      if ((int)MODULE_PAYMENT_SAGEPAY_FORM_ORDER_STATUS_ID > 0) {
        $this->order_status = MODULE_PAYMENT_SAGEPAY_FORM_ORDER_STATUS_ID;
      }

      if (is_object($order)) $this->update_status();

      if (MODULE_PAYMENT_SAGEPAY_FORM_TEST_STATUS == 'true') {
        // normal test mode:
        $this->form_action_url = 'https://test.sagepay.com/vps2form/submit.asp';
        //$this->form_action_url = '../../decrypt_sagepay.php';
        // simulator
        //$this->form_action_url = 'https://test.sagepay.com/Simulator/VSPFormGateway.asp  ';  

      } else {
        $this->form_action_url = 'https://live.sagepay.com/gateway/service/vspform-register.vsp';
      }
    }

// class methods
    function update_status() {
      global $order;

      if ( ($this->enabled == true) && ((int)MODULE_PAYMENT_SAGEPAY_FORM_ZONE > 0) ) {
        $check_flag = false;
        $check_query = tep_db_query("select zone_id from " . TABLE_ZONES_TO_GEO_ZONES . " where geo_zone_id = '" . MODULE_PAYMENT_SAGEPAY_FORM_ZONE . "' and zone_country_id = '" . $order->billing['country']['id'] . "' order by zone_id");
        while ($check = tep_db_fetch_array($check_query)) {
          if ($check['zone_id'] < 1) {
            $check_flag = true;
            break;
          } elseif ($check['zone_id'] == $order->billing['zone_id']) {
            $check_flag = true;
            break;
          }
        }

        if ($check_flag == false) {
          $this->enabled = false;
        }
      }
    }

    function javascript_validation() {
      return false;
    }

    function simpleXor($InString, $Key) {
      $KeyList = array();
      $output = "";

      for($i = 0; $i < strlen($Key); $i++){
        $KeyList[$i] = ord(substr($Key, $i, 1));
      }

      for($i = 0; $i < strlen($InString); $i++) {
        $output.= chr(ord(substr($InString, $i, 1)) ^ ($KeyList[$i % strlen($Key)]));
      }

      return $output;
    }

    function getToken($thisString) {

      $Tokens = array("Status","StatusDetail","VendorTxCode","VPSTxID","TxAuthNo","Amount","AVSCV2");

      $output = array();
      $resultArray = array();

      for ($i = count($Tokens)-1; $i >= 0 ; $i--){
        $start = strpos($thisString, $Tokens[$i]);
        if ($start !== false){
          $resultArray[$i]->start = $start;
          $resultArray[$i]->token = $Tokens[$i];
        }
      }

      sort($resultArray);

      for ($i = 0; $i<count($resultArray); $i++){
        $valueStart = $resultArray[$i]->start + strlen($resultArray[$i]->token) + 1;
        if ($i==(count($resultArray)-1)) {
          $output[$resultArray[$i]->token] = substr($thisString, $valueStart);
        } else {
          $valueLength = $resultArray[$i+1]->start - $resultArray[$i]->start - strlen($resultArray[$i]->token) - 2;
          $output[$resultArray[$i]->token] = substr($thisString, $valueStart, $valueLength);
        }
      }

      return $output;
    }

    function selection() {
      return array('id' => $this->code,
                   'module' => $this->title);
    }

    function pre_confirmation_check() {
      return false;
    }

    function confirmation() {
      return false;
    }

    /** Added by CWS @ theINtoy.co.uk at version 1.18 **/
    function getShoppingCartInfo()
    {
        global $cart, $order;

      $retStr = "";

        if (MODULE_PAYMENT_SAGEPAY_FORM_SHOPCART == 'true') {
    $shipping = $order->info['shipping_cost'];
    $shipping = number_format($shipping, 2, '.', '');
    $products = $cart->get_products();
    $no_lines = sizeof($products);
    $no_lines = $no_lines + 1;

    $shippingStr .= ":Shipping:1:".$shipping.":----:".$shipping.":".$shipping;
    $basketStr = "Basket=".$no_lines;
    $moreStr .= ":More products...See Order.:1:----:----:----:----";
        $moreLen = strlen( $moreStr );
    /*
          Maxmium size of the basket field according to SAGEPAY Form Documentation V2.22
          is 7500 characters. Allowing 7495 as a small safety net.
    */
    $charCount = 7495 - strlen( $shippingStr ) - strlen( $basketStr );

    $detail = "";
        $linesAdded = 0;
        $i = 0;
        $n=sizeof($products);
        $finished = false;

      while ( ( $i<$n ) && ( !$finished ) ) {
          $desc = $products[$i]['name'];
        $desc  = str_replace(":", "", $desc);
        $qty = $products[$i]['quantity'];
        $price = $products[$i]['price'] + $cart->attributes_price($products[$i]['id']);
        $tax = $price/100 * tep_get_tax_rate($products[$i]['tax_class_id']);
        $tax = number_format($tax, 2, '.', '');
        $finalPrice = $price + $tax;
        $finalPrice = number_format($finalPrice, 2, '.', '');
        $lineTotal = $qty * $finalPrice;
        $lineTotal = number_format($lineTotal, 2, '.', '');
        $line = ":".$desc.":".$qty.":".$price.":".$tax.":".$finalPrice.":".$lineTotal;
        $line = str_replace("\n", "", $line);
            $line = str_replace("\r", "", $line);
            $len = strlen( $line );
            if ( ( $charCount - $moreLen - $len) > 0 ) {
                $linesAdded ++;
                $charCount -= $len;
                $detail .= $line;
            } else if ( $charCount - $moreLen > 0 ) {
                $detail .= $moreStr;
                $charCount -= $len;
                $linesAdded ++;
                $finished = true;
            }
            else {
                // We should not hit this point, but if we do lets fininsh
                $finished = true;
            }
            $i++;
      }

      if ( strlen( $detail ) > 0 )
      {
          $linesAdded ++;
          $retStr = "&" . "Basket=" . $linesAdded . $detail . $shippingStr;
      }
      }
      return $retStr;
    }


    function process_button() {
      global $order, $currencies, $currency, $customer_id;

      switch (MODULE_PAYMENT_SAGEPAY_FORM_CURRENCY) {
        case 'Default Currency':
          $sagepay_currency = DEFAULT_CURRENCY;
          break;
        case 'Any Currency':
        default:
          $sagepay_currency = $currency;
          break;
      }

      $plain = "VendorTxCode=" . MODULE_PAYMENT_SAGEPAY_FORM_VENDOR_NAME . date('YmdHis') . $customer_id . "&";
      $plain .= "Amount=" . number_format($order->info['total'] * $currencies->get_value($sagepay_currency), $currencies->get_decimal_places($sagepay_currency), '.', '') . "&";
      $plain .= "Currency=" . $sagepay_currency . "&";
    $plain .= "Description=" . "Goods bought from " . STORE_NAME . "&";
      $plain .= "SuccessURL=" . tep_href_link(FILENAME_CHECKOUT_PROCESS, '', 'SSL', true) . "&";
      $plain .= "FailureURL=" . tep_href_link(FILENAME_CHECKOUT_PROCESS, '', 'SSL', true) . "&";
      $plain .= "CustomerName=" . $order->billing['firstname'] . ' ' . $order->billing['lastname'] . "&";
      $plain .= "CustomerEmail=" . $order->customer['email_address'] . "&";
      $plain .= "VendorEMail=" . STORE_OWNER_EMAIL_ADDRESS . "&";
      
      
      // protocol 2.23 stuff
      $plain .= "BillingSurname=".$order->billing['lastname']."&";
      $plain .= "BillingFirstnames=".$order->billing['firstname']."&";
      $plain .= "BillingCity=".$order->billing['city']."&";
      $plain .= "BillingCountry=".$order->billing['country']['iso_code_2']."&";
      
      $plain .= "DeliverySurname=".$order->delivery['lastname']."&";
      $plain .= "DeliveryFirstnames=".$order->delivery['firstname']."&";
      $plain .= "DeliveryCity=".$order->delivery['city']."&";
      $plain .= "DeliveryCountry=".$order->delivery['country']['iso_code_2']."&";
      // endof 2.23 stuff
      
      
      $delivery_address = $order->delivery['street_address'];
      if (ACCOUNT_SUBURB == 'true') {
        $delivery_address .= ',' . $order->delivery['suburb'];
      }

      $delivery_address .= ',' . $order->delivery['city'];

      if (ACCOUNT_STATE == 'true') {
        $delivery_address .= ',' . $order->delivery['state'];
      }

      $delivery_address .= ',' . $order->delivery['country']['title'];

      $billing_address = $order->billing['street_address'];

      if (ACCOUNT_SUBURB == 'true') {
        $billing_address .= ',' . $order->billing['suburb'];
      }

      $billing_address .= ',' . $order->billing['city'];

      if (ACCOUNT_STATE == 'true') {
        $billing_address .= ',' . $order->billing['state'];
      }

      $billing_address .= ',' . $order->billing['country']['title'];

      $plain .= "DeliveryAddress1=" . $delivery_address . "&";
      $plain .= "DeliveryPostCode=" . $order->delivery['postcode'] . "&";
      $plain .= "BillingAddress1=" . $billing_address . "&";
      $plain .= "BillingPostCode=" . $order->billing['postcode'] . "";

      $cart_string = $this->getShoppingCartInfo();
      $plain .= $cart_string;

      $crypt = base64_encode($this->SimpleXor($plain, MODULE_PAYMENT_SAGEPAY_FORM_PASSWORD));

    if (MODULE_PAYMENT_SAGEPAY_FORM_PREAUTH == 'true') {
        $transaction_type = 'AUTHENTICATE';
    } else {
        $transaction_type = 'PAYMENT';
    }
    
//     $transaction_type = 'REPEAT';
    
      $process_button_string = tep_draw_hidden_field('VPSProtocol', '2.23') .
                               tep_draw_hidden_field('TxType', $transaction_type) .
                               tep_draw_hidden_field('Vendor', MODULE_PAYMENT_SAGEPAY_FORM_VENDOR_NAME) .
                               tep_draw_hidden_field('crypt', $crypt);

      return $process_button_string;
    }


    function before_process() {
      global $HTTP_GET_VARS, $crypt;
      $crypt = $_REQUEST['crypt'];
      $process_button_string = str_replace(" ", "+", $process_button_string);
      $Decoded = $this->SimpleXor(base64_decode(str_replace(" ", "+", $crypt)),MODULE_PAYMENT_SAGEPAY_FORM_PASSWORD);
      $values = $this->getToken($Decoded);

      $StatusDetail = $values['StatusDetail'];

      switch ($values['Status']) {
        case "OK":
        case "AUTHENTICATED":
        case "REGISTERED":
          $Status = true;
        break;

        default:
          $Status = false;
        break;
      }

      if ($Status !== true) {
        $sessionName = tep_session_name();
        $sessionId = $HTTP_GET_VARS[$sessionName];
        $hrefLink = tep_href_link( FILENAME_CHECKOUT_PAYMENT, 'error_message=' . urlencode($StatusDetail), 'SSL', true, false );
        tep_redirect( $hrefLink );
      }

      return $Status;
    }

    function after_process() {
      return false;
    }

    function get_error() {
      global $HTTP_GET_VARS;

      if (isset($HTTP_GET_VARS['message']) && (strlen($HTTP_GET_VARS['message']) > 0)) {
        $error = stripslashes(urldecode($HTTP_GET_VARS['message']));
      } else {
        $error = MODULE_PAYMENT_SAGEPAY_FORM_TEXT_ERROR_MESSAGE;
      }

      return array('title' => MODULE_PAYMENT_SAGEPAY_FORM_TEXT_ERROR,
                   'error' => $error);
    }

    function check() {
      if (!isset($this->_check)) {
        $check_query = tep_db_query("select configuration_value from " . TABLE_CONFIGURATION . " where configuration_key = 'MODULE_PAYMENT_SAGEPAY_FORM_STATUS'");
        $this->_check = tep_db_num_rows($check_query);
      }
      return $this->_check;
    }

    function install() {
      tep_db_query("insert into " . TABLE_CONFIGURATION . " (configuration_title, configuration_key, configuration_value, configuration_description, configuration_group_id, sort_order, set_function, date_added) values ('Enable Protx Form Module', 'MODULE_PAYMENT_SAGEPAY_FORM_STATUS', 'True', 'Do you want to accept Protx Form payments?', '6', '1', 'tep_cfg_select_option(array(\'True\', \'False\'), ', now())");
      tep_db_query("insert into " . TABLE_CONFIGURATION . " (configuration_title, configuration_key, configuration_value, configuration_description, configuration_group_id, sort_order, date_added) values ('Merchant ID', 'MODULE_PAYMENT_SAGEPAY_FORM_VENDOR_NAME', 'TestVendor', 'Vendor Name to use with the Protx Form service', '6', '2', now())");
      tep_db_query("insert into " . TABLE_CONFIGURATION . " (configuration_title, configuration_key, configuration_value, configuration_description, configuration_group_id, sort_order, date_added) values ('Password', 'MODULE_PAYMENT_SAGEPAY_FORM_PASSWORD', 'testvendor', 'Password to use with the Protx Form service', '6', '2', now())");
      tep_db_query("insert into " . TABLE_CONFIGURATION . " (configuration_title, configuration_key, configuration_value, configuration_description, configuration_group_id, sort_order, set_function, date_added) values ('Transaction Currency', 'MODULE_PAYMENT_SAGEPAY_FORM_CURRENCY', 'Any Currency', 'The currency to use for credit card transactions', '6', '3', 'tep_cfg_select_option(array(\'Any Currency\', \'Default Currency\'), ', now())");
      tep_db_query("insert into " . TABLE_CONFIGURATION . " (configuration_title, configuration_key, configuration_value, configuration_description, configuration_group_id, sort_order, date_added) values ('Sort order of display.', 'MODULE_PAYMENT_SAGEPAY_FORM_SORT_ORDER', '0', 'Sort order of display. Lowest is displayed first.', '6', '0', now())");
      tep_db_query("insert into " . TABLE_CONFIGURATION . " (configuration_title, configuration_key, configuration_value, configuration_description, configuration_group_id, sort_order, use_function, set_function, date_added) values ('Payment Zone', 'MODULE_PAYMENT_SAGEPAY_FORM_ZONE', '0', 'If a zone is selected, only enable this payment method for that zone.', '6', '2', 'tep_get_zone_class_title', 'tep_cfg_pull_down_zone_classes(', now())");
      tep_db_query("insert into " . TABLE_CONFIGURATION . " (configuration_title, configuration_key, configuration_value, configuration_description, configuration_group_id, sort_order, set_function, use_function, date_added) values ('Set Order Status', 'MODULE_PAYMENT_SAGEPAY_FORM_ORDER_STATUS_ID', '0', 'Set the status of orders made with this payment module to this value', '6', '0', 'tep_cfg_pull_down_order_statuses(', 'tep_get_order_status_name', now())");
      tep_db_query("insert into " . TABLE_CONFIGURATION . " (configuration_title, configuration_key, configuration_value, configuration_description, configuration_group_id, sort_order, set_function, date_added) values ('Test Mode', 'MODULE_PAYMENT_SAGEPAY_FORM_TEST_STATUS', 'true', 'Use Test Mode?', '6', '4', 'tep_cfg_select_option(array(\'true\', \'false\'), ', now())");
      tep_db_query("insert into " . TABLE_CONFIGURATION . " (configuration_title, configuration_key, configuration_value, configuration_description, configuration_group_id, sort_order, set_function, date_added) values ('Use Pre-Authorisation', 'MODULE_PAYMENT_SAGEPAY_FORM_PREAUTH', 'true', 'Use Pre-Authorisation for all transactions?', '6', '5', 'tep_cfg_select_option(array(\'true\', \'false\'), ', now())");
      tep_db_query("insert into " . TABLE_CONFIGURATION . " (configuration_title, configuration_key, configuration_value, configuration_description, configuration_group_id, sort_order, set_function, date_added) values ('Shopping cart', 'MODULE_PAYMENT_SAGEPAY_FORM_SHOPCART', 'true', 'Send shopping cart details to Protx?', '6', '6', 'tep_cfg_select_option(array(\'true\', \'false\'), ', now())");
    }

    function remove() {
      $keys = '';
      $keys_array = $this->keys();
      for ($i=0; $i<sizeof($keys_array); $i++) {
        $keys .= "'" . $keys_array[$i] . "',";
      }
      $keys = substr($keys, 0, -1);

      tep_db_query("delete from " . TABLE_CONFIGURATION . " where configuration_key in (" . $keys . ")");
    }

    function keys() {
      return array('MODULE_PAYMENT_SAGEPAY_FORM_STATUS', 'MODULE_PAYMENT_SAGEPAY_FORM_VENDOR_NAME', 'MODULE_PAYMENT_SAGEPAY_FORM_PASSWORD', 'MODULE_PAYMENT_SAGEPAY_FORM_CURRENCY', 'MODULE_PAYMENT_SAGEPAY_FORM_ZONE', 'MODULE_PAYMENT_SAGEPAY_FORM_ORDER_STATUS_ID', 'MODULE_PAYMENT_SAGEPAY_FORM_SORT_ORDER', 'MODULE_PAYMENT_SAGEPAY_FORM_TEST_STATUS', 'MODULE_PAYMENT_SAGEPAY_FORM_PREAUTH', 'MODULE_PAYMENT_SAGEPAY_FORM_SHOPCART');
    }
  }
?>
